File: video-parse/src/scene_detector.rs
use image::{DynamicImage, GrayImage};
use anyhow::Result;

/// 场景检测器，用于检测视频中的镜头切换点
pub struct SceneDetector {
    /// 场景变化阈值（像素差异百分比）
    threshold: f64,
    /// 最小场景持续时间（秒）
    min_scene_duration: f64,
}

impl SceneDetector {
    pub fn new(threshold: f64, min_scene_duration: f64) -> Self {
        Self {
            threshold,
            min_scene_duration,
        }
    }

    /// 计算两帧之间的差异度
    /// 返回 0.0-1.0 之间的值，1.0 表示完全不同的帧
    pub fn calculate_frame_difference(&self, frame1: &DynamicImage, frame2: &DynamicImage) -> f64 {
        // 将图像转换为灰度图并缩放到较小尺寸以提高性能
        let gray1 = frame1.to_luma8();
        let gray2 = frame2.to_luma8();
        
        // 计算直方图差异
        let hist_diff = self.calculate_histogram_difference(&gray1, &gray2);
        
        // 计算像素差异
        let pixel_diff = self.calculate_pixel_difference(&gray1, &gray2);
        
        // 组合两种差异度量
        hist_diff * 0.6 + pixel_diff * 0.4
    }

    /// 计算直方图差异
    fn calculate_histogram_difference(&self, img1: &GrayImage, img2: &GrayImage) -> f64 {
        let mut hist1 = [0u32; 256];
        let mut hist2 = [0u32; 256];
        
        for pixel in img1.pixels() {
            hist1[pixel[0] as usize] += 1;
        }
        
        for pixel in img2.pixels() {
            hist2[pixel[0] as usize] += 1;
        }
        
        // 归一化直方图
        let total_pixels = img1.width() * img1.height() as u32;
        let mut diff = 0.0;
        
        for i in 0..256 {
            let h1 = hist1[i] as f64 / total_pixels as f64;
            let h2 = hist2[i] as f64 / total_pixels as f64;
            diff += (h1 - h2).abs();
        }
        
        diff / 2.0 // 归一化到 0-1
    }

    /// 计算像素差异
    fn calculate_pixel_difference(&self, img1: &GrayImage, img2: &GrayImage) -> f64 {
        if img1.width() != img2.width() || img1.height() != img2.height() {
            return 1.0;
        }
        
        let mut diff_sum = 0u64;
        let total_pixels = img1.width() * img1.height();
        
        for (p1, p2) in img1.pixels().zip(img2.pixels()) {
            diff_sum += (p1[0] as i32 - p2[0] as i32).abs() as u64;
        }
        
        diff_sum as f64 / (total_pixels as f64 * 255.0)
    }

    /// 检测场景变化点
    /// 返回场景变化的时间戳（秒）
    pub fn detect_scenes(
        &self,
        frames: &[(f64, DynamicImage)],
        fps: f64,
    ) -> Result<Vec<f64>> {
        if frames.len() < 2 {
            return Ok(vec![0.0]);
        }

        let mut scene_changes = vec![0.0]; // 第一个场景从 0 开始
        let _min_frame_interval = (self.min_scene_duration * fps) as usize;

        for i in 1..frames.len() {
            let prev_idx = i - 1;
            let diff = self.calculate_frame_difference(&frames[prev_idx].1, &frames[i].1);
            
            // 检查是否超过阈值且满足最小时间间隔
            if diff > self.threshold {
                let last_change = *scene_changes.last().unwrap();
                let time_since_last = frames[i].0 - last_change;
                
                if time_since_last >= self.min_scene_duration {
                    scene_changes.push(frames[i].0);
                }
            }
        }

        Ok(scene_changes)
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use image::ImageBuffer;

    #[test]
    fn test_frame_difference() {
        let detector = SceneDetector::new(0.3, 1.0);
        
        // 创建两个不同的图像
        let img1 = DynamicImage::ImageLuma8(
            ImageBuffer::from_fn(100, 100, |_, _| image::Luma([100u8]))
        );
        let img2 = DynamicImage::ImageLuma8(
            ImageBuffer::from_fn(100, 100, |_, _| image::Luma([200u8]))
        );
        
        let diff = detector.calculate_frame_difference(&img1, &img2);
        assert!(diff > 0.0);
    }
}

File: code/video/scene_detector.go
package video

import (
	"fmt"
	"image"
	"os"

	_ "image/jpeg"
)

// SceneDetector 场景检测器
type SceneDetector struct {
	Threshold        float64
	MinSceneDuration float64
}

// NewSceneDetector 创建新的场景检测器
func NewSceneDetector(threshold, minSceneDuration float64) *SceneDetector {
	return &SceneDetector{
		Threshold:        threshold,
		MinSceneDuration: minSceneDuration,
	}
}

// DetectScenes 检测场景变化
func (s *SceneDetector) DetectScenes(frames []FrameInfo, fps float64) ([]float64, error) {
	if len(frames) < 2 {
		return []float64{0.0}, nil
	}

	sceneChanges := []float64{0.0} // 第一个场景从 0 开始

	for i := 1; i < len(frames); i++ {
		prevFrame := frames[i-1]
		currFrame := frames[i]

		// 计算帧差异
		diff, err := s.CalculateFrameDifference(prevFrame.ImagePath, currFrame.ImagePath)
		if err != nil {
			// 如果计算失败，跳过这一帧
			continue
		}

		// 如果差异超过阈值，检测到场景变化
		if diff > s.Threshold {
			// 检查距离上一个场景变化的时间间隔
			timeSinceLastScene := currFrame.Time - sceneChanges[len(sceneChanges)-1]
			if timeSinceLastScene >= s.MinSceneDuration {
				sceneChanges = append(sceneChanges, currFrame.Time)
			}
		}
	}

	return sceneChanges, nil
}

// CalculateFrameDifference 计算两帧之间的差异度
// 返回 0.0-1.0 之间的值，1.0 表示完全不同的帧
func (s *SceneDetector) CalculateFrameDifference(imgPath1, imgPath2 string) (float64, error) {
	// 加载图像
	img1, err := loadImage(imgPath1)
	if err != nil {
		return 0, fmt.Errorf("加载图像1失败: %v", err)
	}

	img2, err := loadImage(imgPath2)
	if err != nil {
		return 0, fmt.Errorf("加载图像2失败: %v", err)
	}

	// 转换为灰度图
	gray1 := toGrayScale(img1)
	gray2 := toGrayScale(img2)

	// 计算直方图差异
	histDiff := s.calculateHistogramDifference(gray1, gray2)

	// 计算像素差异
	pixelDiff := s.calculatePixelDifference(gray1, gray2)

	// 组合差异（直方图差异权重 60%，像素差异权重 40%）
	return histDiff*0.6 + pixelDiff*0.4, nil
}

// loadImage 加载图像
func loadImage(path string) (image.Image, error) {
	file, err := os.Open(path)
	if err != nil {
		return nil, err
	}
	defer file.Close()

	img, _, err := image.Decode(file)
	return img, err
}

// toGrayScale 转换为灰度图
func toGrayScale(img image.Image) *image.Gray {
	bounds := img.Bounds()
	gray := image.NewGray(bounds)

	for y := bounds.Min.Y; y < bounds.Max.Y; y++ {
		for x := bounds.Min.X; x < bounds.Max.X; x++ {
			gray.Set(x, y, img.At(x, y))
		}
	}

	return gray
}

// calculateHistogramDifference 计算直方图差异
func (s *SceneDetector) calculateHistogramDifference(img1, img2 *image.Gray) float64 {
	hist1 := make([]int, 256)
	hist2 := make([]int, 256)

	bounds := img1.Bounds()
	totalPixels := bounds.Dx() * bounds.Dy()

	for y := bounds.Min.Y; y < bounds.Max.Y; y++ {
		for x := bounds.Min.X; x < bounds.Max.X; x++ {
			gray1 := img1.GrayAt(x, y)
			gray2 := img2.GrayAt(x, y)
			hist1[gray1.Y]++
			hist2[gray2.Y]++
		}
	}

	// 归一化并计算差异
	diff := 0.0
	for i := 0; i < 256; i++ {
		h1 := float64(hist1[i]) / float64(totalPixels)
		h2 := float64(hist2[i]) / float64(totalPixels)
		diff += abs(h1 - h2)
	}

	return diff / 2.0 // 归一化到 0-1
}

// calculatePixelDifference 计算像素差异
func (s *SceneDetector) calculatePixelDifference(img1, img2 *image.Gray) float64 {
	bounds := img1.Bounds()
	totalPixels := bounds.Dx() * bounds.Dy()

	diffSum := 0.0
	for y := bounds.Min.Y; y < bounds.Max.Y; y++ {
		for x := bounds.Min.X; x < bounds.Max.X; x++ {
			gray1 := img1.GrayAt(x, y)
			gray2 := img2.GrayAt(x, y)
			diffSum += abs(float64(gray1.Y) - float64(gray2.Y))
		}
	}

	return diffSum / (float64(totalPixels) * 255.0)
}
